src/dotnet/ToolPackage/ToolPackageStore.cs

```diff
@@ -15,13 +15,37 @@ public ToolPackageStore(DirectoryPath root)
 
         public DirectoryPath Root { get; private set; }
 
-        public IEnumerable<IToolPackage> GetInstalledPackages(string packageId)
+        public IEnumerable<IToolPackage> GetInstalledPackages(string packageId = null)
         {
-            if (packageId == null)
+            if (packageId != null)
             {
-                throw new ArgumentNullException(nameof(packageId));
+                return EnumerateVersions(packageId);
             }
 
+            return EnumerateAllPackages().SelectMany(p => p);
+        }
+
+        private IEnumerable<IEnumerable<IToolPackage>> EnumerateAllPackages()
+        {
+            if (!Directory.Exists(Root.Value))
+            {
+                yield break;
+            }
+
+            foreach (var subdirectory in Directory.EnumerateDirectories(Root.Value))
+            {
+                var packageId = Path.GetFileName(subdirectory);
+                if (packageId == ToolPackageInstaller.StagingDirectory)
```

Comment:
Same as what I mentioned. Did we ToLowered or not? it is so hard to tell several calls later. It needs a type.

Reply:
This is coming from the file system so they should already be lowercased.  We should perhaps skip any directory where `ToLowerInvariant` results in a different string from what was enumerated from the directory, though, as a precaution.

Reply:
My point is, the programmer should not need to worry about where it come from and it can be easily done with a specific type. I see the discussion above, we want to use NuGet's external type. That should resolve this problem.

Reply:
I had experience with this kind of external id madness. There was an id that can be represent as hex or dec, and they are all string in the code. Different systems passing with different presentations. Maybe people can track 1 or 2, it got out of control really quick.

Reply:
Completely agree.  Standardizing on the NuGet representation should help here.